x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/hooks/use_completed_cards.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { useKibana } from '../../../../common/lib/kibana';
+import { useStoredCompletedCardIds } from '../../../hooks/use_stored_state';
+import type { OnboardingCardId } from '../../../constants';
+import type {
+  CheckCompleteResult,
+  CheckCompleteResponse,
+  OnboardingGroupConfig,
+  OnboardingCardConfig,
+} from '../../../types';
+import { useOnboardingContext } from '../../onboarding_context';
+
+export type IsCardComplete = (cardId: OnboardingCardId) => boolean;
+export type SetCardComplete = (
+  cardId: OnboardingCardId,
+  completed: boolean,
+  options?: { auto?: boolean }
+) => void;
+export type GetCardCheckCompleteResult = (
+  cardId: OnboardingCardId
+) => CheckCompleteResult | undefined;
+
+export type CardCheckCompleteResult = Partial<Record<OnboardingCardId, CheckCompleteResult>>;
+
+/**
+ * This hook implements the logic for tracking which onboarding cards have been completed using Local Storage.
+ */
+export const useCompletedCards = (cardsGroupConfig: OnboardingGroupConfig[]) => {
+  const { spaceId, reportCardComplete } = useOnboardingContext();
+  const services = useKibana().services;
+
+  // Use stored state to keep localStorage in sync, and a local state to avoid unnecessary re-renders.
+  const [storedCompleteCardIds, setStoredCompleteCardIds] = useStoredCompletedCardIds(spaceId);
+  const [completeCardIds, setCompleteCardIds] = useState<OnboardingCardId[]>(storedCompleteCardIds);
+  // Local state to store the checkCompleteResult for each card
+  const [cardCheckCompleteResult, setCardsCompleteResult] = useState<CardCheckCompleteResult>({});
```

Comment:
nit:

this hook is a little hard to read, is it possible to extract state handling into a reducer with useReducer? if yes it would help a lot and also potentially flatten some setState calls that are invoked here

Reply:
Yes, that would make sense. However, that's not a minor change, and we are very close to FF. I will try to refactor this in a separate PR when this one is merged.